### PR TITLE
feat(repository): add ability to query soft delete records

### PR DESCRIPTION
**Pull request by sumiter92**
## Description

add the ability to query soft delete records

Fixes # (31,18)

## Type of change

- [ ] New feature (non-breaking change which adds functionality)

## Checklist:

- [ ] Performed a self-review of my own code
- [ ] Code conforms with the style guide

### CODE REVIEW

File: src/repositories/soft-crud.repository.base.ts [hunk 1]
```diff
@@ -106,6 +111,45 @@ export abstract class SoftCrudRepository<
     return super.findOne(filter, options);
   }
 
+  //findOne() for soft deleted entry
+  findOneWithSoftDelete(
+    filter?: Filter<T>,
+    options?: Options,
+  ): Promise<(T & Relations) | null> {
+    // Filter out soft deleted entries
+    if (
```
Comment (samarpan-b) [inline src/repositories/soft-crud.repository.base.ts:120]: no need of this logic. Just trigger the original findOne function inside this. That will do find on soft deleted records as well.

**Review by samarpan-b — CHANGES_REQUESTED** [review]
fix comment and update readme please

File: src/repositories/soft-crud.repository.base.ts (hunk 1 above)
Comment (sumiter92) [reply src/repositories/soft-crud.repository.base.ts]: update of the readme is done in the latest commit and make changes in the findOneIncludeSoftDelete function as well.

File: README.md [hunk 2]
```diff
@@ -45,11 +45,11 @@ Right now, this extension exports three abstract classes which are actually help
   If not, then please add these columns to the DB table.
 - **SoftCrudRepository** -
   An abstract base class for all repositories which require soft delete feature.
-  This class is going to be the one which handles soft delete operations and ensures soft deleted entries are not returned in responses at all.
+  This class is going to be the one which handles soft delete operations and ensures soft deleted entries are not returned in responses, However if there is a need to query soft deleted entries as well,there is an options to achieve that and you can use findAll() in place of find() , findOneIncludeSoftDelete() in place of findOne() and findByIdIncludeSoftDelete() in place of of findById(), these will give you the responses including soft deleted entries.
```
Comment (samarpan-b) [inline README.md:48]:
```suggestion
  This class is going to be the one which handles soft delete operations and ensures soft deleted entries are not returned in responses, However if there is a need to query soft deleted entries as well,there is an options to achieve that and you can use findAll() in place of find() , findOneIncludeSoftDelete() in place of findOne() and findByIdIncludeSoftDelete() in place of findById(), these will give you the responses including soft deleted entries.
```

File: README.md [hunk 3]
```diff
@@ -45,11 +45,11 @@ Right now, this extension exports three abstract classes which are actually help
...
   This class is a wrapper over DefaultCrudRepository class from [@loopback/repository](https://github.com/strongloop/loopback-next/tree/master/packages/repository).
 - **DefaultTransactionSoftCrudRepository** -
   An abstract base class for all repositories which require soft delete feature with transaction support.
-  This class is going to be the one which handles soft delete operations and ensures soft deleted entries are not returned in responses at all.
+  This class is going to be the one which handles soft delete operations and ensures soft deleted entries are not returned in responses, However if there is a need to query soft deleted entries as well,there is an options to achieve that and you can use findAll() in place of find() , findOneIncludeSoftDelete() in place of findOne() and findByIdIncludeSoftDelete() in place of of findById(), these will give you the responses including soft deleted entries.
```
Comment (samarpan-b) [inline README.md:52]:
```suggestion
  This class is going to be the one which handles soft delete operations and ensures soft deleted entries are not returned in responses, However if there is a need to query soft deleted entries as well,there is an options to achieve that and you can use findAll() in place of find() , findOneIncludeSoftDelete() in place of findOne() and findByIdIncludeSoftDelete() in place of findById(), these will give you the responses including soft deleted entries.
```

File: src/repositories/soft-crud.repository.base.ts [hunk 4]
```diff
@@ -157,6 +170,24 @@ export abstract class SoftCrudRepository<
     }
   }
 
+  //find by Id including soft deleted record
+  async findByIdIncludeSoftDelete(
+    id: ID,
+    filter?: Filter<T>,
```
Comment (samarpan-b) [inline src/repositories/soft-crud.repository.base.ts:176]: no need of filter here. It doesnot make sense.

**Review by samarpan-b — CHANGES_REQUESTED** [review]

File: src/repositories/soft-crud.repository.base.ts (hunk 4 above)
Comment (sumiter92) [reply src/repositories/soft-crud.repository.base.ts:176]: if anyone wants to use a filter here like a filter excluding where

Comment (samarpan-b) [reply src/repositories/soft-crud.repository.base.ts:176]: Not possible in findbyid

Comment (sumiter92) [reply src/repositories/soft-crud.repository.base.ts:176]: loopback itself gives in its default crud repository something like this **findById(id: ID, filter?: FilterExcludingWhere<T>, options?: Options): Promise<T & Relations>;** here is the link [](https://loopback.io/doc/en/lb4/apidocs.repository.defaultcrudrepository.findbyid.html) and in example too , loopback explains like 
async findById(
    @param.path.number('id') id: number,
    @param.filter(Todo, {exclude: 'where'}) filter?: FilterExcludingWhere<Todo>
  ): Promise<Todo> {
    return this.todoRepository.findById(id, filter);
  }

  @patch('/todos/{id}', {
    responses: {
      '204': {
        description: 'Todo PATCH success',
      },
    },
  })
here is the link to an example resource [](https://loopback.io/doc/en/lb4/Controller-generator.html)

Comment (samarpan-b) [reply src/repositories/soft-crud.repository.base.ts:176]: Ah ! Now I remember why they have it. Its for supporting relations. Makes sense.

File: README.md (hunk 2 above)
Comment (sumiter92) [reply README.md:48]: In Suggested changes, both the deleted and added lines seem the same.

File: README.md (hunk 3 above)
Comment (sumiter92) [reply README.md:52]: In Suggested changes, both the deleted and added lines seem the same.

**Review by samarpan-b — APPROVED** [review]

**Pull request merged**